rosidl_runtime_c/src/string_functions.c

```diff
@@ -171,6 +197,67 @@ rosidl_runtime_c__String__Sequence__fini(
   }
 }
 
+bool
+rosidl_runtime_c__String__Sequence__are_equal(
+  const rosidl_runtime_c__String__Sequence * lhs,
+  const rosidl_runtime_c__String__Sequence * rhs)
+{
+  if (!lhs || !rhs) {
+    return false;
+  }
+  if (lhs->size != rhs->size) {
+    return false;
+  }
+  for (size_t i = 0; i < lhs->size; ++i) {
+    if (!rosidl_runtime_c__String__are_equal(
+        &(lhs->data[i]), &(rhs->data[i])))
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool
+rosidl_runtime_c__String__Sequence__copy(
+  const rosidl_runtime_c__String__Sequence * input,
+  rosidl_runtime_c__String__Sequence * output)
+{
+  if (!input || !output) {
+    return false;
+  }
+  if (output->capacity < input->size) {
+    const size_t allocation_size =
+      input->size * sizeof(rosidl_runtime_c__String);
+    rosidl_runtime_c__String * data =
+      (rosidl_runtime_c__String *)realloc(output->data, allocation_size);
+    if (!data) {
+      return false;
+    }
+    for (size_t i = output->capacity; i < input->size; ++i) {
+      if (!rosidl_runtime_c__String__init(&data[i])) {
+        /* free currently allocated and return false */
+        for (; i-- > output->capacity; ) {
+          rosidl_runtime_c__String__fini(&data[i]);
+        }
+        free(data);
+        return false;
+      }
+    }
+    output->data = data;
+    output->size = input->size;
```

Comment:
In cases where the output capacity is greater than the input size, the output size is not set because this line is not executed.

It seems to me that this is a bug. There are two possible ways to fix it:
- Set the output size unconditionally
- Set the output size to the input size when the input size is larger.

Reply:
> It seems to me that this is a bug. 

It is. 

> Set the output size unconditionally

That gets us the closest to a regular copy constructor.

---

Would you be willing to contribute such a fix @nnmm ?

Reply:
Sure. I just pushed a new commit to the backport PRs and will make a new PR for master.